### ISSUE TITLE
CA1508 - False positive for null check in finally block

### ISSUE THREAD

**Issue by bartdesmet** [dotnet/roslyn-analyzers#4672]
```csharp
static void Repro()
{
    D d = null;
    try
    {
        d = new D();
    }
    finally
    {
        d?.Dispose();
    }
}

class D : IDisposable
{
    public D() => throw new InvalidOperationException();

    public void Dispose() { }
}
```

The constructor of `D` can throw and leave `d` to be `null`, so the `finally` block correctly checks for `null`. However, CA1508 suggests to remove the null check.

**Comment by Youssef1313** [comment]
@bartdesmet Because there is no `catch` clause, the above program should crash and the finally block won't be reachable with `null` value.

**Comment by Youssef1313** [comment]
Ok I now see that if `Repro` is called in a `try` clause that has a `catch` the null check will be making a difference.

**Comment by Youssef1313** [comment]
@mavasani Should the following check be done regardless it's compiler generated or not?

https://github.com/dotnet/roslyn-analyzers/blob/60fec5c28143441c2311e14c77f032a7c21cc19b/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidDeadConditionalCode.cs#L155-L156

**Comment by mavasani** [comment]
>  or will this introduce false negatives?

Yes, that is correct. We cannot assume that first block of finally cannot have redundant null checks. For example, if the original repro is adjusted a bit such that `d` is guaranteed to be non-null prior to try block, `d?.Dispose()` indeed has a redundant `?.` which needs to be flagged.

This issue likely needs further debugging through the dataflow analysis for try/catch/finally blocks in https://github.com/dotnet/roslyn-analyzers/blob/7edc7c3925c5b66a513bf6adecbc07504b8e9351/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowAnalysis.cs#L173-L185

**Comment by skrysmanski** [comment]
@bartdesmet 

> Because there is no `catch` clause, the above program should crash and the finally block won't be reachable with `null` value.

I think this is wrong. The `finally` block is always executed - even on exceptions (that's what `finally` is for).

**Comment by skrysmanski** [comment]
On a different note: I'm getting CA1508 for this code:

```c#
X509Certificate2? pubCert = null;

try
{
    pubCert = new X509Certificate2(publicKeyBytes);

    if (pubCert.HasPrivateKey)
    {
        throw new InvalidOperationException("This certificate already has a private key.");
    }
}
catch (Exception)
{
    pubCert?.Dispose(); // <-- CA1508: 'pubCert' is never 'null'.
    throw;
}
```

This is wrong because the constructor of `X509Certificate2` can throw exceptions in which case `pubCert` would be `null`.

I'm guessing this is the same problem as in this issue? Or should I create a separate issue for this?

**Comment by genlu** [comment]
I think it's likely the same issue, no need to create a separate one for now.

**Comment by mavasani** [comment]
Fixed with #6828, fix should be in the upcoming .NET8 preview SDK.
